litestar/contrib/pydantic/utils.py

```diff
@@ -0,0 +1,75 @@
+from __future__ import annotations
+
+from pydantic import VERSION as PYDANTIC_VERSION
+from typing_extensions import Any, get_type_hints
+
+from litestar.utils.predicates import is_generic
+from litestar.utils.typing import (
+    _substitute_typevars,
+    _zip,
+    get_origin_or_inner_type,
+    get_type_hints_with_generics_resolved,
+    instantiable_type_mapping,
+    unwrap_annotation,
+    wrapper_type_set,
+)
+
+if PYDANTIC_VERSION.startswith("2"):
+    # These utility functions are the same as the ones in litestar.utils.typing but this has to be
+    # done due to this issue: https://github.com/pydantic/pydantic/issues/7837. This is only for
+    # v2.
+
+    def pydantic_unwrap_and_get_origin(annotation: Any) -> Any | None:
```

Comment:
We could save the duplication by making the callable that gets the origin an argument to `get_origin_or_inner_type()` in `utils.typing` without any impact on existing functionality, e.g,:

```py 
def get_origin_or_inner_type(annotation: T, get_origin_callable: Callable[[Any], Any] = get_origin) -> Any:
    origin = get_origin_callable(annotation)
    ...
```

Then could do something like:

```py
if PYDANTIC_VERSION.startswith("2"):
    get_origin_or_inner_type = partial(get_origin_or_inner_type(get_origin_callable=special_get_origin_for_pydantic_v2)
```

Reply:
I do like this, but should we add this configurability to handle solely a bug with Pydantic. Imo, the `get_origin` call should return the right origin even for third-party libraries and if it doesn't it's a bug on their side? The reason I'm a bit reluctant is because I don't see a reason for using that `get_origin_callable` anywhere else.

Also, if and when Pydantic does fix this issue, then we could just remove all this Pydantic specific code and just use the original ones in `utils.typing`.

Reply:
Just to be clear, I'm fine either way, but I think these are points we should consider before making the decision.

Reply:
Your point is valid - I did consider this and can see both sides. You are right, it is a shame to have to modify the method to support this, but at the same time, the modification really makes no functional difference to the current behaviour.

What about making the parameters to the functions private with leading underscore and undocumented so we can remove them freely if/when pydantic repair it.

Or, leave it - I'm OK with that too 👍